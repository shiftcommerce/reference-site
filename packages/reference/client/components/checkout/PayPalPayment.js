// Libraries
import { Component } from 'react'

class PayPalPayment extends Component {
  render () {
    return <div className='o-form__wrapper'>
      <h3>PayPal</h3>
    </div>
  }
}

export default PayPalPayment
