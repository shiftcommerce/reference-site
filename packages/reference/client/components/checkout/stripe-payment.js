// Libraries
import { Component } from 'react'

// Components
import AddressBook from '../address-book'
import StripeWrapper from './stripe-wrapper'

// Objects
import { Button, Checkbox, CheckoutAddressForm } from 'shift-react-components'

class StripePayment extends Component {
  renderFormSubmitButton () {
    return (
      <div className='o-form__input-group'>
        <Button
          aria-label='Review Your Order'
          className='c-address-form__button o-button--sml'
          label='Review Your Order'
          status={(this.props.nextStepAvailable() ? 'positive' : 'disabled')}
          type='primary'
          disabled={!this.props.nextStepAvailable()}
          onClick={() => this.props.nextSection('complete')}
        />
      </div>
    )
  }

  render () {
    const {
      autoFillAddress,
      billingAddress,
      billingAsShipping,
      cardTokenRequested,
      changeBillingAsShipping,
      checkout,
      countries,
      currentAddress,
      onBlur,
      onChange,
      onShowField,
      onCardTokenReceived,
      setCardErrors,
      shippingAddress
    } = this.props

    return (
      <>
        <div className='o-form__background'>
          <div className='o-form__wrapper'>
            <StripeWrapper
              cardTokenRequested={cardTokenRequested}
              onCardTokenReceived={onCardTokenReceived}
              setCardErrors={setCardErrors}
              checkout={checkout}
            />

            <div className='o-form__input-group'>
              <label>Billing address *</label>
            </div>

            <Checkbox
              type='checkbox'
              label='Same as shipping address'
              name='shippingAddressAsBillingAddress'
              checked={billingAsShipping}
              onChange={changeBillingAsShipping}
            />

            { billingAsShipping &&
              <div aria-label='Shipping address to be used for billing' className='o-payment-method__address-summary  c-payment-method__address-summary'>
                <p className='u-bold'>{ `${shippingAddress.first_name} ${shippingAddress.last_name}` }</p>
                <p>{ shippingAddress.address_line_1 }</p>
                <p>{ shippingAddress.address_line_2 }</p>
                <p>{ shippingAddress.city }</p>
                <p>{ shippingAddress.postcode }</p>
              </div>
            }
          </div>
        </div>

        { !billingAsShipping &&
          <>
            { !this.props.addressBookEmpty() && <AddressBook
              formName='shippingAddress'
              currentAddressId={billingAddress.id}
              onNewAddress={this.props.onNewAddress}
              onBookAddressSelected={this.props.onBookAddressSelected}
              addressFormDisplayed={this.props.addressFormDisplayed()}
            /> }
            { this.props.addressFormDisplayed() && <CheckoutAddressForm
              aria-label='Billing address form'
              autoFillAddress={autoFillAddress}
              checkout={checkout}
              countries={countries}
              currentAddress={currentAddress}
              title='Billing Address'
              formName='billingAddress'
              onChange={onChange}
              onBlur={onBlur}
              className='o-form__billing'
              onShowField={onShowField}
            /> }
          </>
        }
        { this.renderFormSubmitButton() }
      </>
    )
  }
}

export default StripePayment
