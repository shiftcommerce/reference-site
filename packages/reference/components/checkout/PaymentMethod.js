// Libraries
import { Component } from 'react'

// Components
import AddressForm from './AddressForm'
import CheckboxInputGroup from './CheckboxInputGroup'

// Actions
import { toggleCollapsed } from '../../actions/checkoutActions'

export default class PaymentMethod extends Component {
  onToggleCollapsed () {
    return (
      this.props.dispatch(toggleCollapsed(this.props.formName))
    )
  }

  constructor () {
    super()
    this.onToggleCollapsed = this.onToggleCollapsed.bind(this)
  }

  renderSecurityFields () {
    // Example data
    const months = [
      '', '01', '02', '03', '04', '05', '06',
      '07', '08', '09', '10', '11', '12'
    ]

    const years = [ '', '17', '18', '19', '20', '21', '22' ]

    const monthDropdown = months.map((month, index) =>
      <option value={parseInt(month)} key={index}>
        {month}
      </option>
    )

    const yearDropdown = years.map((year, index) =>
      <option value={parseInt(year)} key={index}>
        {year}
      </option>
    )

    return <div className='o-payment-method__security-fields'>
      <label className='o-payment-method__label' aria-label='Card expiry date' htmlFor='expiry_month'>Expiration *</label>
      <select className='o-payment-method__input' aria-label='Expiry month' id='expiry_month' required='required'>
        {monthDropdown}
      </select>
      <span className='o-payment-method__text'>
        /
      </span>

      <select className='o-payment-method__input' aria-label='Expiry year' id='expiry_year' required='required'>
        {yearDropdown}
      </select>

      <label className='o-payment-method__label-cvv' aria-label='CVV security code' htmlFor='cvv'>
        CVV *
        <span className='o-payment-method__text'>Hint</span>
      </label>
      <input className='o-payment-method__input' type='text' id='cvv' required='required' />
    </div>
  }

  render () {
    const {checkout, formName} = this.props
    const paymentMethod = checkout.paymentMethod
    const shippingAddress = checkout.shippingAddress
    const billingAddress = paymentMethod.shippingAddressAsBillingAddress ? shippingAddress : checkout.billingAddress
    const collapsed = paymentMethod.collapsed

    return <div aria-label='Payment method' className='o-form'>
      <h3>Payment Method</h3>

      {collapsed &&
        <div>
          <button aria-label='Edit payment method' className='c-button' onClick={this.onToggleCollapsed}>Edit</button>

          <div aria-label='Payment method summary' className='o-form__wrapper'>
            <p>
              <span>VISA **** **** **** 1607 </span>
              <span>Exp: 01/24</span>
            </p>
            <p>
              <span className='u-bold'>Billing address: {billingAddress.fullName}, </span>
              <span>{billingAddress.address1}, {billingAddress.city}, {billingAddress.postCode}</span>
            </p>
          </div>
        </div>
      }

      {!collapsed &&
        <div>
          <button aria-label='Use credit or debit card' className='c-button'>Credit / Debit card</button>
          <button aria-label='Use paypal' className='c-button c-button--disabled' disabled='disabled'>Paypal</button>

          <div className='o-form__wrapper'>
            <form>
              <div className='o-form__input-group'>
                <label aria-label='Credit or debit card number' htmlFor='card_number'>Credit / Debit Card Number *</label>
                <input type='text' id='card_number' required='required' />
              </div>

              {this.renderSecurityFields()}

              <div className='o-form__input-group'>
                <label>Billing address *</label>
              </div>

              <CheckboxInputGroup
                fieldLabel='Same as shipping address'
                fieldName='shippingAddressAsBillingAddress'
                fieldValue={paymentMethod.shippingAddressAsBillingAddress}
                formName={formName}
                dispatch={this.props.dispatch}
              />
            </form>

            {paymentMethod.shippingAddressAsBillingAddress &&
              <div aria-label='Shipping address to be used for billing' className='o-payment-method__address-summary'>
                <h4>{shippingAddress.fullName}</h4>
                <p>{shippingAddress.address1}</p>
                <p>{shippingAddress.address2}</p>
                <p>{shippingAddress.city}</p>
                <p>{shippingAddress.postCode}</p>
              </div>
            }

            {!(paymentMethod.shippingAddressAsBillingAddress) &&
              <AddressForm aria-label='Billing address form' checkout={this.props.checkout} title='Billing Address' formName='billingAddress' addressType='billing' dispatch={this.props.dispatch} />
            }
          </div>
        </div>
      }

      <div aria-label='Use gift card or rewards code' className='o-form__wrapper'>
        <p>
          + Use a Gift Card / Rewards Code
        </p>
      </div>
    </div>
  }
}
