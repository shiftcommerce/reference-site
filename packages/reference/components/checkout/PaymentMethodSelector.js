// Libraries
import { Component } from 'react'

// Objects
import Image from '../../objects/Image'

export default class PaymentMethodSelector extends Component {
  buttonClass (paymentType) {
    return `c-payment-method__button ${this.props.currentMethod === paymentType ? 'c-payment-method__button--active' : ''}`
  }

  renderPaymentMethodOption (label, ariaLabel, paymentMethod) {
    const { onPaymentMethodChanged } = this.props
    return (
      <div aria-label={ariaLabel} className={this.buttonClass(paymentMethod)} onClick={() => onPaymentMethodChanged(paymentMethod)}>
        <span>{ label }</span>
      </div>
    )
  }

  render () {
    return <>
      <div className='o-form c-payment-method'>
        { this.renderPaymentMethodOption('Credit/Debit Card', 'Use credit or debit card', 'card') }
        { this.renderPaymentMethodOption('PayPal', 'Use paypal', 'paypal') }
      </div>
      <div className='c-payment-method__images'>
        <span className='c-payment-method__images-cardpayment'>
          <Image src='/static/payments/VISA.svg' />
          <Image src='/static/payments/Mastercard.svg' />
          <Image src='/static/payments/american-express.svg' />
          <Image src='/static/payments/maestro.svg' />
        </span>
        <span className='c-payment-method__images-paypal'>
          <Image src='/static/payments/PayPal.svg' />
        </span>
      </div>
    </>
  }
}
