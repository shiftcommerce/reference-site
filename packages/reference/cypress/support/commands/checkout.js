/**
 * Adds variant to cart and proceeds to checkout
 * @param {integer} variantId - the variant ID
 * @param {integer} quantity - the required quantity
 */
Cypress.Commands.add('addVariantToCartAndProceedToCheckout', () => {
  // Setup server
  cy.server()
  // Stub add to cart request
  cy.route({
    method: 'POST',
    url: '/addToCart',
    status: 200,
    response: 'fixture:cart/add-clock-computer.json'
  }).as('addToCart')
  // Add multiple variants to cart
  cy.addVariantsToCart()
  // visit cart page
  cy.contains(/View Shopping Basket/i).click()
  // Click the checkout button
  cy.get('.c-cart-summary-buttons__cta--proceed').click()
})

/**
 * Navigates to the shipping method checkout page
 */
Cypress.Commands.add('navigateToShippingMethodCheckoutPage', () => {
  // Setup server
  cy.server()

  // Uses custom command - Cypress/support/commands/empty-search.js
  cy.emptySearch()

  // Stub get cart request
  cy.route({
    method: 'GET',
    url: '/getCart',
    status: 200,
    response: 'fixture:cart/get-cart-with-line-item.json'
  }).as('getCart')

  // Stub create address request
  cy.route({
    method: 'POST',
    url: '/createAddress',
    status: 201,
    response: 'fixture:checkout/paypal/create-address.json'
  }).as('createAddress')

  // Stub get shipping methods request
  cy.route({
    method: 'GET',
    url: '/getShippingMethods',
    status: 200,
    response: 'fixture:cart/get-shipping-methods.json'
  }).as('getShippingMethods')

  // Stub set cart billing address request
  cy.route({
    method: 'POST',
    url: '/setCartBillingAddress',
    status: 201,
    response: 'fixture:checkout/paypal/set-cart-billing-address.json'
  }).as('setCartBillingAddress')

  // Stub set cart shipping address request
  cy.route({
    method: 'POST',
    url: '/setCartShippingAddress',
    status: 201,
    response: 'fixture:checkout/paypal/set-cart-shipping-address.json'
  }).as('setCartShippingAddress')

  // Stub set shipping method request
  cy.route({
    method: 'POST',
    url: '/setShippingMethod',
    status: 200,
    response: 'fixture:checkout/paypal/set-shipping-method.json'
  }).as('setShippingMethod')

  // Add item to cart and proceed to checkout
  cy.addVariantToCartAndProceedToCheckout()
  // Check we are on the Payment Method Page
  cy.url().should('includes', '/checkout/payment-method')
  // Wait for PayPal script and page to be loaded
  cy.wait(2500)
  // Click the test PayPal btn
  cy.contains(/Test PayPal Button/i).click()
  // Check its the Shipping Methods Page
  cy.url().should('includes', '/checkout/shipping-method')
})

/**
 * Navigates to the review checkout page
 */
Cypress.Commands.add('navigateToReviewCheckoutPage', () => {
  // Setup server
  cy.server()

  // Stub patch PayPal order request
  cy.route({
    method: 'POST',
    url: '/patchPayPalOrder',
    status: 200,
    response: {}
  }).as('patchPayPalOrder')

  // Navigate to shipping method checkout page
  cy.navigateToShippingMethodCheckoutPage()
  // Wait for page to fully load
  cy.wait(2000)
  // Navigate to the Review Page
  cy.contains(/Review Your Order/i).click()
  // Check patch PayPal order request has been made
  cy.wait('@patchPayPalOrder')
  // Check its the Review Page Page
  cy.url().should('includes', '/checkout/review')
  // Check Payment Details section header is present
  cy.get('.c-payment__header').contains(/Payment Details/i)
})

/**
 * Stubs the authorize PayPal order request
 */
Cypress.Commands.add('stubSuccessfulAuthorizePayPalOrderRequest', () => {
  // Setup server
  cy.server()

  // Stub authorize PayPal order request
  cy.route('POST', '/authorizePayPalOrder', {
    status: 201,
    data: {
      // generate random id
      id: `${Math.random().toString(36).substring(2) + (new Date()).getTime().toString(36)}`,
      // the returned PayPal status
      status: 'COMPLETED',
      // expires a day from now
      expirationTime: new Date(new Date().getTime() + 1000 * 60 * 60 * 24)
    }
  }).as('authorizePayPalOrder')
})
