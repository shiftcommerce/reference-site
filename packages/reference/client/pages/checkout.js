// Libraries
import { Component } from 'react'
import { connect } from 'react-redux'
import Router from 'next/router'

// Libs
import InputFieldValidator from '../lib/input-field-validator'

// Actions
import { readCheckoutFromLocalStorage,
  toggleCollapsed,
  setShippingMethod,
  changeBillingAddress,
  setValidationMessage,
  changePaymentMethod,
  inputChange,
  inputComplete,
  showField } from '../actions/checkout-actions'
import { readCart, initializeCart, updateQuantity } from '../actions/cart-actions'
import {
  createOrder,
  requestCardToken,
  setCardToken,
  setPaymentError,
  setCardErrors
} from '../actions/order-actions'

// Objects
import Logo from '../objects/logo'

// Components
import CustomHead from '../components/custom-head'
import CheckoutSteps from '../components/checkout/checkout-steps'
import AddressForm from '../components/checkout/address-form'
import CheckoutCart from '../components/checkout/checkout-cart'
import CheckoutCartTotal from '../components/checkout/checkout-cart-total'
import ShippingMethods from '../components/checkout/shipping-methods'
import PaymentMethod from '../components/checkout/payment-method'
import PromoInput from '../components/promo-input'

export class CheckoutPage extends Component {
  static async getInitialProps ({ reduxStore }) {
    return {}
  }
  constructor () {
    super()

    this.onToggleCollapsed = this.onToggleCollapsed.bind(this)
    this.nextSection = this.nextSection.bind(this)
    this.setShippingMethod = this.setShippingMethod.bind(this)
    this.onInputChange = this.onInputChange.bind(this)
    this.onInputBlur = this.onInputBlur.bind(this)
    this.onShowField = this.onShowField.bind(this)
    this.changeBillingAddress = this.changeBillingAddress.bind(this)
    this.convertToOrder = this.convertToOrder.bind(this)
    this.onPaymentMethodChanged = this.onPaymentMethodChanged.bind(this)
    this.onCardTokenReceived = this.onCardTokenReceived.bind(this)
    this.setCardErrors = this.setCardErrors.bind(this)
    this.updateQuantity = this.updateQuantity.bind(this)
    this.deleteItem = this.deleteItem.bind(this)
  }

  componentDidMount () {
    this.props.dispatch(readCheckoutFromLocalStorage())
    this.props.dispatch(readCart()).then((cart) => {
      if (cart.totalQuantity === 0) Router.push('/cart')
    })
  }

  componentWillReceiveProps (props) {
    const prevOrder = this.props.order
    const newOrder = props.order
    // Only when new order got created,
    // Redirect to order confirmation page
    if (newOrder && newOrder.id && newOrder.id !== prevOrder.id) {
      Router.push('/order')
      props.dispatch(initializeCart())
    }
  }

  updateQuantity (e) {
    const lineItem = {
      sku: e.target.dataset.variant,
      quantity: parseInt(e.target.value, 10)
    }
    this.props.dispatch(updateQuantity(lineItem))
  }

  deleteItem (e) {
    e.preventDefault()
    const lineItem = {
      sku: e.target.dataset.variant,
      quantity: 0
    }
    this.props.dispatch(updateQuantity(lineItem))
    if (this.props.cart.lineItems.length === 0) {
      Router.push('/cart')
    }
  }

  nextSection (eventType) {
    const { checkout } = this.props

    let componentName

    if (checkout.currentStep === 1) {
      componentName = 'shippingAddress'
    } else if (checkout.currentStep === 2) {
      componentName = 'shippingMethod'
    } else if (checkout.currentStep === 3) {
      componentName = 'paymentMethod'
    }

    window.scrollTo(0, 0)
    this.props.dispatch(toggleCollapsed(eventType, componentName))
  }

  onToggleCollapsed (eventType, componentName) {
    window.scrollTo(0, 0)
    this.props.dispatch(toggleCollapsed(eventType, componentName))
  }

  setShippingMethod (shippingMethod) {
    this.props.dispatch(setShippingMethod(shippingMethod))
  }

  onInputChange (event, formName, fieldName, fieldValue) {
    this.props.dispatch(inputChange(formName, fieldName, fieldValue))
  }

  validateInput (formName, fieldName, fieldValue, rules) {
    let validationMessage = new InputFieldValidator(fieldName, fieldValue, rules).validate()
    this.props.dispatch(setValidationMessage(formName, fieldName, validationMessage))
  }

  onInputBlur (event, formName, fieldName, fieldValue, rules) {
    this.validateInput(formName, fieldName, fieldValue, rules)
    this.props.dispatch(inputComplete())
  }

  convertToOrder (event) {
    if (this.props.checkout.paymentMethod.selectedMethod === 'card') {
      this.props.dispatch(requestCardToken(true))
    } else {
      this.props.dispatch(createOrder(this.props.cart, this.props.checkout, this.props.order))
    }
  }

  onShowField (formName, fieldName) {
    this.props.dispatch(showField(formName, fieldName))
  }

  onPaymentMethodChanged (paymentMethod) {
    this.props.dispatch(changePaymentMethod(paymentMethod))
  }

  onCardTokenReceived ({ error, token }) {
    if (error) {
      this.props.dispatch(setPaymentError(error.message))
    } else {
      this.props.dispatch(setCardToken(token))
    }
    this.props.dispatch(requestCardToken(false))
  }

  setCardErrors (error) {
    this.props.dispatch(setCardErrors(error))
  }

  changeBillingAddress (event, formName, fieldName) {
    this.props.dispatch(changeBillingAddress(formName, fieldName, event.target.checked))
  }

  isPaymentValid () {
    let billingAddressErrors = this.props.checkout.billingAddress.errors
    const isCardValid = !this.props.order.cardErrors
    const isBillingAddressValid = Object.keys(billingAddressErrors).every((key) => billingAddressErrors[key] === '') === true
    return isBillingAddressValid && isCardValid
  }

  render () {
    const { checkout, cart } = this.props
    const hasLineItems = cart.totalQuantity > 0
    if (checkout.loading) {
      return (
        <div>
          <CustomHead />
          loading...
        </div>
      )
    } else if (checkout.error) {
      return (
        <p>{ checkout.error }</p>
      )
    } else {
      return (
        <>
          <CustomHead />
          {hasLineItems &&
            <>
              <div className='o-header'>
                <div className='c-step-indicators__logo'>
                  <Logo className='u-text-color--primary' />
                </div>
                <div>
                  <CheckoutSteps {...this.props} />
                </div>
              </div>
              <div className='o-grid-container'>
                <div className='o-col-1-13 o-col-1-9-l'>
                  <AddressForm {...this.props}
                    title='Shipping Address'
                    formName='shippingAddress'
                    addressType='shipping'
                    onChange={this.onInputChange}
                    onBlur={this.onInputBlur}
                    onShowField={this.onShowField}
                    onToggleCollapsed={this.onToggleCollapsed}
                  />
                  <ShippingMethods
                    {...this.props}
                    formName='shippingMethod'
                    setShippingMethod={this.setShippingMethod}
                    onToggleCollapsed={this.onToggleCollapsed}
                  />
                  <PaymentMethod
                    formName='paymentMethod'
                    {...this.props}
                    changeBillingAddress={this.changeBillingAddress}
                    onChange={this.onInputChange}
                    onBlur={this.onInputBlur}
                    onPaymentMethodChanged={this.onPaymentMethodChanged}
                    onToggleCollapsed={this.onToggleCollapsed}
                    onCardTokenReceived={this.onCardTokenReceived}
                    setCardErrors={this.setCardErrors}
                  />
                </div>
                <PromoInput />
                <div className='o-col-1-13 o-col-9-13-l c-cart-table'>
                  <CheckoutCart title='Your Cart' {...this.props} updateQuantity={this.updateQuantity} deleteItem={this.deleteItem} />
                  <CheckoutCartTotal {...this.props} convertToOrder={this.convertToOrder} onClick={() => { this.nextSection('complete') }} />
                </div>
              </div>
            </>
          }
        </>
      )
    }
  }
}

const mapStateToProps = (state) => {
  return {
    checkout: state.checkout || {},
    cart: state.cart || {},
    order: state.order || {},
    ip: state.ip
  }
}

const mapDispatchToProps = (dispatch) => {
  return {
    dispatch: dispatch
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(CheckoutPage)
