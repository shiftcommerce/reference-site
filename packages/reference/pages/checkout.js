// Libraries
import { Component } from 'react'
import withRedux from 'next-redux-wrapper'
import Router from 'next/router'
import Link from 'next/link'

// Utils
import { configureStore } from '../utils/configureStore'

// Actions
import { readCheckoutFromLocalStorage,
  toggleCollapsed,
  setShippingMethod,
  setShippingBillingAddress,
  inputChange,
  inputComplete,
  showField } from '../actions/checkoutActions'
import { readCart } from '../actions/cartActions'

// Objects
import Image from '../objects/Image'

// Components
import CustomHead from '../components/CustomHead'
import CheckoutSteps from '../components/checkout/CheckoutSteps.js'
import AddressForm from '../components/checkout/AddressForm.js'
import CheckoutCart from '../components/checkout/CheckoutCart.js'
import CheckoutCartTotal from '../components/checkout/CheckoutCartTotal.js'
import ShippingMethods from '../components/checkout/ShippingMethods.js'
import PaymentMethod from '../components/checkout/PaymentMethod.js'

export class CheckoutPage extends Component {
  constructor (props) {
    super(props)

    this.onToggleCollapsed = this.onToggleCollapsed.bind(this)
    this.setShippingMethod = this.setShippingMethod.bind(this)
    this.onInputChange = this.onInputChange.bind(this)
    this.onInputBlur = this.onInputBlur.bind(this)
    this.onShowField = this.onShowField.bind(this)
    this.setBillingShippingAddress = this.setBillingShippingAddress.bind(this)
  }

  componentDidMount () {
    this.props.dispatch(readCheckoutFromLocalStorage())
    this.props.dispatch(readCart())

    if (this.props.cart.lineItems.length === 0) Router.push('/cart')
  }

  onToggleCollapsed (componentName) {
    this.props.dispatch(toggleCollapsed(componentName))
  }

  setShippingMethod (shippingMethod) {
    this.props.dispatch(setShippingMethod(shippingMethod))
  }

  onInputChange (event, formName, fieldName) {
    this.props.dispatch(inputChange(formName, fieldName, event.target.value))
  }

  onInputBlur () {
    this.props.dispatch(inputComplete())
  }

  onShowField (formName, fieldName) {
    this.props.dispatch(showField(formName, fieldName))
  }

  setBillingShippingAddress (event, formName, fieldName) {
    this.props.dispatch(setShippingBillingAddress(formName, fieldName, event.target.checked))
  }

  render () {
    const { checkout, cart } = this.props
    const hasLineItems = cart.lineItems.length > 0
    const paymentMethodCollapsed = checkout.paymentMethod.collapsed

    return (
      <div>
        <CustomHead />
        {hasLineItems &&
          <div>
            <div className='o-header'>
              <span className='o-header__logo'>
                <Link href='/home/index' as='/'>
                  <a>
                    <Image width={200} height={80} />
                  </a>
                </Link>
              </span>
              <CheckoutSteps {...this.props} />
            </div>
            <div className="o-grid-container">
              <div className="o-col-1-13 o-col-1-9-l">
                <AddressForm {...this.props}
                  title='Shipping Address'
                  formName='shippingAddress'
                  addressType='shipping'
                  onChange={this.onInputChange}
                  onBlur={this.onInputBlur}
                  onShowField={this.onShowField}
                  onToggleCollapsed={this.onToggleCollapsed}
                />
                <ShippingMethods {...this.props} formName='shippingMethod' setShippingMethod={this.setShippingMethod} />
                <PaymentMethod
                  formName='paymentMethod'
                  {...this.props}
                  setBillingShippingAddress={this.setBillingShippingAddress}
                  onChange={this.onInputChange}
                  onBlur={this.onInputBlur}
                />
                {!paymentMethodCollapsed &&
                  <div className='o-form'>
                    <button className='c-button' onClick={() => { this.onToggleCollapsed('paymentMethod') }}>Review Your Order</button>
                  </div>
                }
              </div>
              <div className="o-col-1-13 o-col-9-13-l">
                <CheckoutCart title='Your Cart' {...this.props} />
                <CheckoutCartTotal {...this.props} />
              </div>
            </div>
          </div>
        }
      </div>
    )
  }
}

const mapStateToProps = (state) => {
  return {
    checkout: state.checkout || {},
    cart: state.cart || {}
  }
}

const mapDispatchToProps = (dispatch) => {
  return {
    dispatch: dispatch
  }
}

export default withRedux(configureStore, mapStateToProps, mapDispatchToProps)(CheckoutPage)
