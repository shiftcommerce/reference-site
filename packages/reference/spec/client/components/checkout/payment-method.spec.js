// Components
import PaymentMethod from '../../../../client/components/checkout/payment-method'
import PaymentMethodHeader from '../../../../client/components/checkout/payment-method-header'
import PaymentMethodSelector from '../../../../client/components/checkout/payment-method-selector'
import PayPalPayment from '../../../../client/components/checkout/paypal-payment'
import '../../../../client/components/checkout/stripe-payment'

// Mock Stripe payment
jest.mock('../../../../client/components/checkout/stripe-payment', () => (props) => <p>Stripe payment</p>)

test('renders the payment method header and selector', () => {
  const wrapper = shallow(<PaymentMethod />)

  expect(wrapper).toMatchSnapshot()
  expect(wrapper).toContainReact(<PaymentMethodHeader />)
  expect(wrapper).toContainReact(<PaymentMethodSelector />)
})

test('renders StripePayment when card payment is selected', () => {
  const wrapper = mount(<PaymentMethod selectedPaymentMethod={'card'} cart={{}} order={{}} />)

  expect(wrapper).toMatchSnapshot()
  expect(wrapper).toContainReact(<p>Stripe payment</p>)
})

test('renders PayPalPayment when card payment is selected', () => {
  const wrapper = mount(<PaymentMethod selectedPaymentMethod={'paypal'} cart={{}} order={{}} />)

  expect(wrapper).toMatchSnapshot()
  expect(wrapper).toContainReact(<PayPalPayment />)
})
