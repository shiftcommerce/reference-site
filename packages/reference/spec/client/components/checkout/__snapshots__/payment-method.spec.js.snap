// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders PayPalPayment when card payment is selected 1`] = `
<PaymentMethod
  cart={Object {}}
  order={Object {}}
  selectedPaymentMethod="paypal"
>
  <div
    aria-label="Payment method"
    className="o-form  c-payment-method"
  >
    <PaymentMethodHeader>
      <div
        className="o-form__header  c-payment-method__header"
      >
        <h2>
          Payment Method
        </h2>
      </div>
    </PaymentMethodHeader>
    <div
      className="c-payment-method__section"
      style={
        Object {
          "display": "block",
        }
      }
    >
      <PaymentMethodSelector
        currentMethod="paypal"
      >
        <div
          className="o-form c-payment-method"
        >
          <div
            className="c-payment-method__button-group"
          >
            <Button
              aria-label="Use credit or debit card"
              className="c-payment-method__button "
              label="Credit/Debit Card"
              onClick={[Function]}
            >
              <button
                aria-label="Use credit or debit card"
                className="c-payment-method__button  o-button"
                disabled={false}
                onClick={[Function]}
                role="button"
                tabIndex="0"
              >
                Credit/Debit Card
              </button>
            </Button>
            <Button
              aria-label="Use paypal"
              className="c-payment-method__button c-payment-method__button--active"
              label="PayPal"
              onClick={[Function]}
            >
              <button
                aria-label="Use paypal"
                className="c-payment-method__button c-payment-method__button--active o-button"
                disabled={false}
                onClick={[Function]}
                role="button"
                tabIndex="0"
              >
                PayPal
              </button>
            </Button>
          </div>
          <div
            className="c-payment-method__images"
          >
            <span
              className="c-payment-method__images-cardpayment"
            >
              <Image
                src="/static/payments/visa.svg"
              >
                <div>
                  <picture>
                    <img
                      className="o-image"
                      src="/static/payments/visa.svg"
                      srcSet="/static/payments/visa.svg 769w"
                    />
                  </picture>
                </div>
              </Image>
              <Image
                src="/static/payments/mastercard.svg"
              >
                <div>
                  <picture>
                    <img
                      className="o-image"
                      src="/static/payments/mastercard.svg"
                      srcSet="/static/payments/mastercard.svg 769w"
                    />
                  </picture>
                </div>
              </Image>
              <Image
                src="/static/payments/american-express.svg"
              >
                <div>
                  <picture>
                    <img
                      className="o-image"
                      src="/static/payments/american-express.svg"
                      srcSet="/static/payments/american-express.svg 769w"
                    />
                  </picture>
                </div>
              </Image>
              <Image
                src="/static/payments/maestro.svg"
              >
                <div>
                  <picture>
                    <img
                      className="o-image"
                      src="/static/payments/maestro.svg"
                      srcSet="/static/payments/maestro.svg 769w"
                    />
                  </picture>
                </div>
              </Image>
            </span>
            <span
              className="c-payment-method__images-paypal"
            >
              <Image
                src="/static/payments/pay-pal.svg"
              >
                <div>
                  <picture>
                    <img
                      className="o-image"
                      src="/static/payments/pay-pal.svg"
                      srcSet="/static/payments/pay-pal.svg 769w"
                    />
                  </picture>
                </div>
              </Image>
            </span>
          </div>
        </div>
      </PaymentMethodSelector>
      <PayPalPayment>
        <div
          className="o-form__wrapper"
        >
          <h3>
            PayPal
          </h3>
        </div>
      </PayPalPayment>
    </div>
  </div>
</PaymentMethod>
`;

exports[`renders StripePayment when card payment is selected 1`] = `
<PaymentMethod
  cart={Object {}}
  order={Object {}}
  selectedPaymentMethod="card"
>
  <div
    aria-label="Payment method"
    className="o-form  c-payment-method"
  >
    <PaymentMethodHeader>
      <div
        className="o-form__header  c-payment-method__header"
      >
        <h2>
          Payment Method
        </h2>
      </div>
    </PaymentMethodHeader>
    <div
      className="c-payment-method__section"
      style={
        Object {
          "display": "block",
        }
      }
    >
      <PaymentMethodSelector
        currentMethod="card"
      >
        <div
          className="o-form c-payment-method"
        >
          <div
            className="c-payment-method__button-group"
          >
            <Button
              aria-label="Use credit or debit card"
              className="c-payment-method__button c-payment-method__button--active"
              label="Credit/Debit Card"
              onClick={[Function]}
            >
              <button
                aria-label="Use credit or debit card"
                className="c-payment-method__button c-payment-method__button--active o-button"
                disabled={false}
                onClick={[Function]}
                role="button"
                tabIndex="0"
              >
                Credit/Debit Card
              </button>
            </Button>
            <Button
              aria-label="Use paypal"
              className="c-payment-method__button "
              label="PayPal"
              onClick={[Function]}
            >
              <button
                aria-label="Use paypal"
                className="c-payment-method__button  o-button"
                disabled={false}
                onClick={[Function]}
                role="button"
                tabIndex="0"
              >
                PayPal
              </button>
            </Button>
          </div>
          <div
            className="c-payment-method__images"
          >
            <span
              className="c-payment-method__images-cardpayment"
            >
              <Image
                src="/static/payments/visa.svg"
              >
                <div>
                  <picture>
                    <img
                      className="o-image"
                      src="/static/payments/visa.svg"
                      srcSet="/static/payments/visa.svg 769w"
                    />
                  </picture>
                </div>
              </Image>
              <Image
                src="/static/payments/mastercard.svg"
              >
                <div>
                  <picture>
                    <img
                      className="o-image"
                      src="/static/payments/mastercard.svg"
                      srcSet="/static/payments/mastercard.svg 769w"
                    />
                  </picture>
                </div>
              </Image>
              <Image
                src="/static/payments/american-express.svg"
              >
                <div>
                  <picture>
                    <img
                      className="o-image"
                      src="/static/payments/american-express.svg"
                      srcSet="/static/payments/american-express.svg 769w"
                    />
                  </picture>
                </div>
              </Image>
              <Image
                src="/static/payments/maestro.svg"
              >
                <div>
                  <picture>
                    <img
                      className="o-image"
                      src="/static/payments/maestro.svg"
                      srcSet="/static/payments/maestro.svg 769w"
                    />
                  </picture>
                </div>
              </Image>
            </span>
            <span
              className="c-payment-method__images-paypal"
            >
              <Image
                src="/static/payments/pay-pal.svg"
              >
                <div>
                  <picture>
                    <img
                      className="o-image"
                      src="/static/payments/pay-pal.svg"
                      srcSet="/static/payments/pay-pal.svg 769w"
                    />
                  </picture>
                </div>
              </Image>
            </span>
          </div>
        </div>
      </PaymentMethodSelector>
      <Component>
        <p>
          Stripe payment
        </p>
      </Component>
    </div>
  </div>
</PaymentMethod>
`;

exports[`renders the payment method header and selector 1`] = `
<div
  aria-label="Payment method"
  className="o-form  c-payment-method"
>
  <PaymentMethodHeader />
  <div
    className="c-payment-method__section"
    style={
      Object {
        "display": "block",
      }
    }
  >
    <PaymentMethodSelector />
  </div>
</div>
`;
